File: Codeforces/1288/1288c.cpp
#include<bits/stdc++.h>
using namespace std;

int main()
{
	long long n, m;
	cin >> n >> m;
	long long mod = 1000000007;
	int i, j, k, x, y;
	vector<vector<vector<long long>>> dparr;
	if(m == 1)
	{
		cout << n*(n+1)/2 << "\n";
	}
	else {
	for(x=1; x<=n; x++)
	{
		vector<vector<long long>> dp;

		vector<long long> temp;
		for(i=0; i<x; i++)
			temp.push_back(1);

		dp.push_back(temp);
		for(i=1; i<m; i++)
		{
			vector<long long> lol;
			long long su = 0;
			for(j=0; j<x; j++)
			{
				su += dp[i-1][j];
				lol.push_back(su);
				su %= mod;
			}
			
			dp.push_back(lol);
			// cout << i << "\n";
		}
		dparr.push_back(dp);
		// cout <<"AA";
	}

	long long ans = 0;
	for(i=1; i<n+1; i++)
	{
		for(j=i; j<n+1; j++)
		{
			long long su1 = 0;
			for(k=0; k<dparr[i-1][m-2].size(); k++)
			{
				su1 += dparr[i-1][m-2][k];
				
			}
			su1 %= mod;
			
			long long su2 = 0;
			for(k=0; k<dparr[n-j][m-2].size(); k++)
			{
				su2 += dparr[n-j][m-2][k];
				
			}
			su2 %= mod;
			
			ans += (su1*su2)%mod;
			
		}
		ans %= mod;
	}
	cout << ans << "\n";}
}

File: Codeforces/1288/1288c.py
import copy

n, m = map(int, input().split())
dp = [[0 for i in range(n)] for j in range(m)]
# print(dp)
for i in range(n):
	dp[0][i] = n-i
mod = 10**9 + 7

# dp1 = [[0 for i in range(n)] for j in range(m)]

# for i in range(n):
# 	dp1[0][i] = 1

# for i in range(1, m):
# 	for j in range(n):
# 		dp1[i][j] = sum(dp1[i-1][:j+1])

# print(dp1)

dparr = []
for x in range(1, n+1):
	dp1 = [[0 for i in range(x)] for j in range(m)]

	for i in range(x):
		dp1[0][i] = 1

	for i in range(1, m):
		for j in range(x):
			dp1[i][j] = sum(dp1[i-1][:j+1])

	dparr += [copy.deepcopy(dp1)]
	# print(dp1)
# for i in range(1, m):
# 	for j in range(n):
# 		su = 0
# 		for k in range(n-j):
# 			su += dp[i-1][k]
# 			su %= mod
# 		dp[i][j] = su

# print(dp[:3])


# print(sum(dp[m-1])%mod)
su = 0
for i in range(1, n+1):
	for j in range(i, n+1):
		lol = sum(dparr[i-1][m-2])*sum(dparr[n-j][m-2])
		# print(i, j, lol)
		su += (sum(dparr[i-1][m-2])*sum(dparr[n-j][m-2]))%mod
		su %= mod

print(su)


